added solution for Data Structures and Unit Testing

IhorStetskyi opened the pull request:
Data Structures and Unit Tests HWs are added

IhorStetskyi committed:
added solution for Data Structures and Unit Testing

IhorStetskyi committed:
Merge branch 'main' into Resolved_HW_6_and_7(DataStructures_and_UnitTesting)

IhorStetskyi committed:
.gitignore is now working2

Hekitkoo commented on Stetskyi_Homework_6/DataStructures/DataStructures/Tasks/DoublyLinkedList.cs:
```diff
@@ -5,43 +5,118 @@
 namespace Tasks
 {
-    public class DoublyLinkedList<T> : IDoublyLinkedList<T>
+    public class DoublyLinkedList<T> : IDoublyLinkedList<T>, IEnumerator<T>, IEnumerator
```
@IhorStetskyi, there is no need IEnumerator. Now IEnumerator<T> already inherited IEnumerator.

IhorStetskyi replied:
IEnumerator removed

Hekitkoo commented on Stetskyi_Homework_6/DataStructures/DataStructures/Tasks/DoublyLinkedList.cs:
```diff
@@ -5,43 +5,118 @@
+    public class DoublyLinkedList<T> : IDoublyLinkedList<T>, IEnumerator<T>, IEnumerator
     {
-        public int Length => throw new NotImplementedException();
+        public T[] elementArray = new T[0];
```
@IhorStetskyi, Array.Empty<T>(); more clear variant. Change it, please.

Hekitkoo replied:
@IhorStetskyi, also, no need for 'elementArray' to be public.

IhorStetskyi replied:
elementArray initialized with Array.Empty<T>()
element array is private now

Hekitkoo commented on Stetskyi_Homework_6/DataStructures/DataStructures/Tasks/DoublyLinkedList.cs:
```diff
@@ -5,43 +5,118 @@
+                if (element.Equals(item) && !removed)
+                {
+                    removed = true;
+                    continue;
```
@IhorStetskyi, no need to continue the algorithm if were found a needed element to delete.

IhorStetskyi replied:
In this case, I think we have to because it helps us to skip one result.Add iteration. In this type of deletion, we will not have an empty element after deletion and our list length will be decreased by 1.
Please correct me if I am getting something wrong

<img width="309" alt="remove_Continue" src="https://user-images.githubusercontent.com/105840950/170110043-fb402bb7-1ec5-4eb6-b817-a6d84054c4ea.png">

Hekitkoo commented on Stetskyi_Homework_6/DataStructures/DataStructures/Tasks/HybridFlowProcessor.cs:
```diff
@@ -5,24 +5,34 @@ namespace Tasks
         public T Dequeue()
         {
-            throw new NotImplementedException();
+            if (source.Length == 0)
```
@IhorStetskyi, extract validation to a method, please.

IhorStetskyi replied:
Extracted validation to SourceNotEmptyValidation() method

Hekitkoo commented on Stetskyi_Homework_7/Unit Testing/ImplementedKatas/FizzBuzz/FizzBuzz/Program.cs:
```diff
@@ -0,0 +1,51 @@
+
+        public static string WriteFizzBuzz(int myint)
+        {
+            ExceptionHandler(myint);
```
@IhorStetskyi, rename a method, please. (Use verb when write methods)

IhorStetskyi replied:
WriteFizzBuzz renamed to IterateThroughArrayAndDisplayResult

Hekitkoo commented on Stetskyi_Homework_7/Unit Testing/ImplementedKatas/FizzBuzz/FizzBuzz/Program.cs:
```diff
@@ -0,0 +1,51 @@
+            }
+            else if (myint % 5 == 0)
+            {
+                return "Buzz";
```
@IhorStetskyi, extract to local const, please.

IhorStetskyi replied:
numbers 3 & 5 extracted to local consts multiplyToThree & multiplyToFive

Hekitkoo commented on Stetskyi_Homework_7/Unit Testing/ImplementedKatas/FizzBuzz/FizzBuzzUnitTests/UnitTest1.cs:
```diff
@@ -0,0 +1,123 @@
+    {
+
+        [SetUp]
+        public void Setup()
```
@IhorStetskyi, if it is empty you could remove it.

IhorStetskyi replied:
SetUp removed

Hekitkoo commented on Stetskyi_Homework_7/Unit Testing/ImplementedKatas/FizzBuzz/FizzBuzzUnitTests/UnitTest1.cs:
```diff
@@ -0,0 +1,123 @@
+
+namespace FizzBuzzUnitTests
+{
+    public class Tests
```
@IhorStetskyi, your filename UnitTest1 while the class is named Tests. Please apply equal naming.

IhorStetskyi replied:
Tests class name is changed to UnitTest1

Hekitkoo commented on Stetskyi_Homework_7/Unit Testing/ImplementedKatas/DictionaryReplacer/DictionaryReplacer/ExtensionClass.cs:
```diff
@@ -0,0 +1,64 @@
+
+namespace DictionaryReplacer
+{
+    public static class ExtensionClass
```
@IhorStetskyi, there are many extensions in the 'ExtensioClass'. 'Extensions' will be a more clear name. Apply it, please.

IhorStetskyi replied:
ExtensionClass renamed to Extensions

Hekitkoo commented on Stetskyi_Homework_7/Unit Testing/ImplementedKatas/CalcStats/CalcStats/Program.cs:
```diff
@@ -0,0 +1,25 @@
+            }
+
+            StatsCalculator statsCalculator = new StatsCalculator();
+            ResultClass result = new ResultClass(statsCalculator.FindMin(array), statsCalculator.FindMax(array), statsCalculator.GetCount(array), statsCalculator.GetAvg(array));
```
@IhorStetskyi, I believe we could make this line cleaner.
1. or -> Add to ResultClass (no need to write 'Class', by the way) method that incapsulate Calculation;
2. or -> create variables above and pass them to class.

IhorStetskyi replied:
Renamed ResultClass to Result
Added calculation logic to Result class
Added one more constructor that takes an array as a parameter

Hekitkoo commented on Stetskyi_Homework_7/Unit Testing/ImplementedKatas/CalcStats/CalcStats/StatsCalculator.cs:
```diff
@@ -0,0 +1,50 @@
+{
+    public class StatsCalculator
+    {
+
```
@IhorStetskyi, an empty line, remove it, please.

IhorStetskyi replied:
removed empty line

Hekitkoo commented on Stetskyi_Homework_7/Unit Testing/ImplementedKatas/CalcStats/CalcStats/StatsCalculator.cs:
```diff
@@ -0,0 +1,50 @@
+            }
+            return res;
+        }
+        public decimal GetAvg(params int[] myparams)
```
@IhorStetskyi, add an empty line between methods, please.

IhorStetskyi replied:
strange, I can not see any empty lines between methods

<img width="284" alt="no empty lines" src="https://user-images.githubusercontent.com/105840950/170111051-4b679c93-bfb0-4989-909a-657f76261829.png">

Hekitkoo reviewed (changes requested):
@IhorStetskyi, I like how you make the test task. Nice job!

IhorStetskyi committed:
Class: DataStructures/DataStructures/Tasks/DoublyLinkedList.cs
IEnumerator removed
elementArray initialized with Array.Empty<T>()
element array is private now

Class: DataStructures/DataStructures/Tasks/HybridFlowProcessor.cs
Extracted validation to SourceNotEmptyValidation() method

Class: FizzBuzz/Program.cs
WriteFizzBuzz renamed to IterateThroughArrayAndDisplayResult
numbers 3 & 5 extracted to local consts multiplyToThree & multiplyToFive

Class: FizzBuzzUnitTests/UnitTest1.cs
SetUp removed
Tests class name is changed to UnitTest1

Class: DictionaryReplacer/ExtensionClass.cs
ExtensionClass renamed to Extensions

Class: CalcStats/Program.cs
Renamed ResultClass to Result
Added calculation logic to Result class
Added one more constructor that takes an array as a paramether

Class: CalcStats/StatsCalculator.cs
removed empty line
